Fix printing of attributes on polymorphic variant leafs

IwanKaramazow opened the pull request:
Fix for https://github.com/facebook/reason/issues/821

example:
```reason
external readFileSync :
  name::string =>
  [ | `utf8 | `my_name [@bs.as "ascii"]]
  [@bs.string] =>
  string =
  "" [@@bs.module "fs"];

external readFileSync2 :
  name::string =>
  [
    | `utf8 [@bs.as "ascii"]
    | `my_name [@bs.as "ascii"]
  ]
  [@bs.string] =>
  string =
  "" [@@bs.module "fs"];
```

yunxing commented on src/reason_pprint_ast.ml:
```diff
@@ -2597,9 +2597,21 @@ class printer  ()= object(self:'self)
+            [ | `utf8 | `my_name [@bs.as "ascii"]]
+            type variantType = | Foo int [@onInt]
+        We don't want [@bs.as "ascii"], nor [@onInt] dropped. Because of [@bs.as "ascii"]
+        coming in from the Ast nodes outside of the Parsetree.constructor_declaration, while
```
Not sure if the comment is true: I think both of the attributes are inside of contructor_declaration.  The difference is that in the first example, the type is an abstract type and all Rtags are defined in the manifest section of constructor_declaration. In the second example, the attribute [@onInt]  is decorating a variant type, which is also defined in constructor_delcaration.

IwanKaramazow replied:
You're right, I misinterpreted the whole thing. Thanks for pointing this out! Is the overall approach ok here?

IwanKaramazow commented:
Has been rebased on master.

IwanKaramazow commented:
@yunxing just rebased this one again on master.
Should I improve something in this PR? Someone ran into this problem this week & mentioned it on Discord.

jordwalke commented:
Thanks for the rebase onto master! Let's merge this by EOD? Ping me if it's not in by then?

IwanKaramazow commented:
Ping @jordwalke ☺️

jordwalke commented on reason-parser/src/reason_pprint_ast.ml:
```diff
@@ -2618,9 +2618,10 @@ class printer  ()= object(self:'self)
       | (_::_, None) -> add_bar nameOf (normalize args)
       | (_::_, Some res) -> add_bar nameOf (normalize (args@[res]))
     in
+    let allAttrs = List.concat [stdAttrs; attrs_on_leaf] in
```
Why weren't the attrs_on_leaf part of the standard attributes to begin with?

jordwalke commented on reason-parser/src/reason_pprint_ast.ml:
```diff
@@ -2566,12 +2566,12 @@ class printer  ()= object(self:'self)
   (* TODOATTRIBUTES: Attributes on the entire variant leaf are likely
    * not parsed or printed correctly. *)
-  method type_variant_leaf1 opt_ampersand polymorphic print_bar x =
+  method type_variant_leaf1 attrs_on_leaf opt_ampersand polymorphic print_bar x =
```
I haven't looked at this code in a while, so my memory is rusty here, but how is it that we have attributes that are on the "entire leaf", joined with the "standard attributes", which are all printed as equals, and then later when we parse we somehow can recover which was which?

IwanKaramazow commented:
@jordwalke 
So the problem was:
```
       | Ptyp_variant (l, closed, low) ->
          let pcd_loc = x.ptyp_loc in
          let pcd_attributes = x.ptyp_attributes in
          let pcd_res = None in
          let variant_helper rf =
            match rf with
              | Rtag (label, attrs, opt_ampersand, ctl) ->
                let pcd_name = {
                  txt = label;
                  loc = pcd_loc;
                } in
                let pcd_args = Pcstr_tuple ctl in
                let all_attrs = List.concat [pcd_attributes; attrs] in
                self#type_variant_leaf ~opt_ampersand ~polymorphic:true {pcd_name; pcd_args; pcd_res; pcd_loc; pcd_attributes = all_attrs}
```
If you look add `| Rtag (label, attrs, opt_ampersand, ctl)`, it has attributes which fell outside of the `pcd_attributes` because they're in the `Rtag`. I just modified the PR to drop the weird logic I introduced and just join the `'Rtag attrs'` with the normal `pcd_attributes`. This way our printer handles them correctly.
